### DISCUSSION TITLE
Mailer specific default_url_options

### DISCUSSION THREAD

**Discussion by thedumbtechguy** [janko/rodauth-rails#306] [General]
Hi. As always, awesome work and thank you. Loving it so much.

Alright. So right now, we support url generation using Action mailer's `default_url_options`.

https://github.com/janko/rodauth-rails/blob/9d2daf934e00ead48a144fa9557cbdced1fe35f2/lib/rodauth/rails/feature/internal_request.rb#L47

This works great in general. However, we have a multi subdomain I attempted to use the [rails per mailer overrides](https://api.rubyonrails.org/classes/ActionMailer/Base.html#class-ActionMailer::Base-label-Generating+URLs).

However, it appears from the above shared code, this always defaults to the global config.

It would be nice to have some way to pass the url options in while defaulting to the global config.
Happy to do the work if you describe how you'd like something like this, **if** you'd like something like this, implemented.

**Comment by thedumbtechguy** [comment]
This is similar and I see overlaps in how passing in options could be extended to support this

https://github.com/janko/rodauth-rails/discussions/296

**Comment by thedumbtechguy** [comment]
I also see that this override is supported in controllers

https://guides.rubyonrails.org/v3.2/action_controller_overview.html#default_url_options

**Comment by janko** [comment]
Thanks for the suggestion, I didn't know you can set `default_url_options` per mailer. I initially wanted to suggest to just override Rodauth's `domain` directly, but I see value in not duplicating configuration that probably needs to be in the mailer anyway.

Maybe adding `#rails_url_options` to the configuration DSL would suffice?

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    rails_url_options RodauthMailer.default_url_options
  end
end
```

If that seems reasonable to you, feel free to send a PR 🙂

**Reply by thedumbtechguy** [reply]
So in my mind, I was thinking maybe a way to pass it in when you create the rodauth instance?

That would allow controller and mailer overrides since they'd be passing the options at last moment when it's fully resolved.

Unfortunately, I'm not exactly sure how to go about doing that.

**Reply by janko** [reply]
Aha, you're asking because `default_url_options` are overridden in mailers and controllers at the instance level, not the class level. Most likely it won't be possible to add it to the constructor, because that's already controlled by Rodauth. I will think about this more 🤔 

A current workaround could be adding an attribute accessor:

```rb
class RodauthMain < Rodauth::Rails::Auth
  configure do
    # ...
  end

  attr_accessor :default_url_options

  private

  def rails_url_options
    default_url_options || super
  end
end
```

Then in the mailer:

```rb
class RodauthMailer < ApplicationMailer
  # ...
  private

  def rodauth(name, account_id, &block)
    instance = RodauthApp.rodauth(name).allocate
    instance.instance_eval { @account = account_ds(account_id).first! }
    instance.default_url_options = default_url_options
    instance.instance_eval(&block) if block
    instance
  end
end
```

For the actual solution, maybe we could make `#rails_url_options` public and have it accept a hash:

```rb
instance.rails_url_options(default_url_options)
```

Maybe it's weird to have a method double as getter and setter, just thinking out loud.

**Reply by thedumbtechguy** [reply]
The proposal with "attribute accessor" works for what I need.

I'll pursue the workaround in my app and see how it goes.

**Reply by thedumbtechguy** [reply]
This solution works excellently.

**Reply by janko** [reply]
I just pushed a change to main that allows modifying `rodauth.rails_url_options`:

```rb
instance.rails_url_options.merge!(default_url_options)
```

I find this conceptually a simpler than a new attribute accessor or a getter/setter method.
